lex/lexer.ml

```diff
@@ -0,0 +1,306 @@
+        || (('\248' <= first) && (first <= '\255'))
+  then begin
+          let continue = ref true in
+          let i = ref 1 in
+          while !continue do
+            (* ['A'-'Z' 'a'-'z' '_' '\192'-'\214' '\216'-'\246' '\248'-'\255' '\'' '0'-'9'] *)
+            let chr = Bytes.get buff !i in
+            if (('A' <= chr) && (chr <= 'Z')) || (('a' <= chr) && (chr <= 'z'))
+                || (chr = '_') || (('\192' <= chr) && (chr <= '\214'))
+                || (('\216' <= chr) && (chr <= '\246'))
+                || (('\248' <= chr) && (chr <= '\255')) || (chr = '\'')
+                || (('0' <= chr) && (chr <= '9'))
+            then incr i
+            else continue := false
+          done;
+          advance lexbuf !i;
+          Bytes.to_string (Bytes.sub buff 0 !i)
+  end else raise Bad_rule
+
+let decimal_code  c d u =
+  100 * (Char.code c - 48) + 10 * (Char.code d - 48) + (Char.code u - 48)
+
+let lex_char lexbuf =
+  let buff = get_next_buff lexbuf 6 in
+  if (Bytes.length buff) = 0 then raise Bad_rule;
+  let first = Bytes.get buff 0 in
+  if first = '\''
+  then begin
+    match Bytes.get buff 1 with
+    | '\\' -> begin match Bytes.get buff 2 with
+              | '\\' -> begin advance lexbuf 4; Tchar (Char.code('\\')) end
+              | '\'' -> begin advance lexbuf 4; Tchar (Char.code('\'')) end
+              | '"' ->  begin advance lexbuf 4; Tchar (Char.code('"')) end
+              | ' ' ->  begin advance lexbuf 4; Tchar (Char.code(' ')) end
+              | 'r' ->  begin advance lexbuf 4; Tchar (Char.code('\013')) end
+              | 'b' ->  begin advance lexbuf 4; Tchar (Char.code('\008')) end
+              | 'n' ->  begin advance lexbuf 4; Tchar (Char.code('\010')) end
+              | 't' ->  begin advance lexbuf 4; Tchar (Char.code('\009')) end
+              (* We should check that we actually have that many chars left,
+               * and that they are numbers. Fortunately, lexer.mll is
+               * well-formed. *)
+              | _ ->    begin advance lexbuf 6;
+                Tchar (decimal_code (Bytes.get buff 2) (Bytes.get buff 3)
+                                    (Bytes.get buff 4)) end
+              end
+    | c -> if (Bytes.get buff 2) = '\''
+           then begin advance lexbuf 3; Tchar (Char.code(c))
+           end else raise Bad_rule
+  end else raise Bad_rule
+
+
+let skip_space_and_comments lexbuf =
+  let str = get_next_buff lexbuf 2 in
+  match (Bytes.get str 0) with
+  | '(' -> if (Bytes.get str 1) = '*'
+           then begin
+             advance lexbuf 2;
+             lex_comment lexbuf;
+           end else
+             raise Bad_rule
+  | ' ' -> advance lexbuf 1
+  | '\009' -> advance lexbuf 1
+  | '\010' -> advance lexbuf 1
+  | '\012' -> advance lexbuf 1
+  | '\013' -> advance lexbuf 1
+  | _ -> raise Bad_rule
+
+let rec try_skip_space_and_comments lexbuf =
+  try
+  skip_space_and_comments lexbuf;
+  try_skip_space_and_comments lexbuf
+  with _ -> ();;
```

Comment:
It is nicer to write the function as follows, which makes it tail-recursive. (With the current definition the recursive call is within a `try..with` handler, so not in tail position)

```ocaml
let rec try_skip_space_and_comments lexbuf =
  match skip_space_and_comments lexbuf with
  | () -> try_skip_space_and_comments lexbuf
  | exception Bad_rule -> ()
```